elastic/kibana
# [Connector] API for adding OAuth support to ServiceNow connectors

Towards https://github.com/elastic/kibana/issues/130306

NOTE: This is targeted at a feature branch

## Summary

Addresses items 1-5 as described in https://github.com/elastic/kibana/issues/130306. I will open a separate PR for item 6.

1. Extends `config` and `secrets` schema for all ServiceNow connectors to hold values needed for OAuth
  * `config` - Added `isOAuth`, `clientID`, `jwtKeyId` and `userIdentifierValue` to config. If `isOAuth` is `true`, these fields must be populated. These fields are `schema.nullable` in order to support a connector switching from Basic Auth to OAuth and back.
  * `secrets` - Added `clientSecret`, `privateKey` and `privateKeyPassword` to secrets. If `isOAuth` is true, these fields must be populated. These fields are `schema.nullable` in order to support a connector switching from Basic Auth to OAuth and back.
2. Added validation for all newly added fields. Also added connector validation to ensure that the correct `secrets` configuration is specified based on the `isOAuth` config.
3. Refactored creation of `axios` instance inside the service now executors. Previously, if the connector type was `.servicenow-itom` or `.servicenow-sir`, it looks like 2 axios instances were created. Refactored to remove this duplication and extracted out the creation of the axios instance to either use `username` and `password` for Basic Auth or use an axios interceptor if OAuth fields are specified.
4. Added logic to store the access token as a `connector_token` saved object. This is the same logic as in the email connector, so refactored a little to reduce duplicate code.
5. Added migration function to add `isOAuth: false` to the config for all existing ServiceNow connectors.

## To Verify

You will need to set up ServiceNow as described in the Google Doc to retrieve the values required for `clientId`, `clientSecret`, `jwtKeyId`, `privateKey` and `privateKeyPassword`

1. Verify that you can still create all 3 service now connector types using the API with basic auth credentials.

<details>
<summary>Sample Requests</summary>


ServiceNow ITSM Basic Auth
```
POST https://localhost:5601/api/actions/connector
{
  "name": "Test Servicenow Basic Auth API",
  "connector_type_id": ".servicenow",
  "config": {
    "apiUrl": "https://dev102869.service-now.com",
    "usesTableApi": false
  },
  "secrets": {
    "username": <username>,
    "password": <password>
  }
}
```

ServiceNow SIR Basic Auth
```
POST https://localhost:5601/api/actions/connector
{
  "name": "Test Servicenow SIR Basic Auth API",
  "connector_type_id": ".servicenow-sir",
  "config": {
    "apiUrl": "https://dev102869.service-now.com",
    "usesTableApi": false
  },
  "secrets": {
    "username": <username>,
    "password": <password>
  }
}
```

ServiceNow ITOM Basic Auth
```
POST https://localhost:5601/api/actions/connector
{
  "name": "Test Servicenow ITOM Basic Auth API",
  "connector_type_id": ".servicenow-itom",
  "config": {
    "apiUrl": "https://dev102869.service-now.com"
  },
  "secrets": {
    "username": <username>,
    "password": <password>
  }
}
```
</details>

2. Verify that you can create all 3 service now connector types using the API with OAuth credentials with privateKeyPassword

<details>
<summary>Sample Requests</summary>

ServiceNow ITSM OAuth
```
POST https://localhost:5601/api/actions/connector
{
  "name": "Test Servicenow OAuth API",
  "connector_type_id": ".servicenow",
  "config": {
    "apiUrl": "https://dev102869.service-now.com",
    "usesTableApi": false,
    "isOAuth": true,
    "clientId": <clientId>,
    "userIdentifierValue": <email>,
    "jwtKeyId": <jwtKeyId>
  },
  "secrets": {
    "clientSecret": <clientSecret>,
    "privateKey": <privateKey>,
    "privateKeyPassword": <privateKeyPassword>
  }
}
```

ServiceNow SIR OAuth
```
POST https://localhost:5601/api/actions/connector
{
  "name": "Test Servicenow SIR OAuth API",
  "connector_type_id": ".servicenow-sir",
  "config": {
    "apiUrl": "https://dev102869.service-now.com",
    "usesTableApi": false,
    "isOAuth": true,
    "clientId": <clientId>,
    "userIdentifierValue": <email>,
    "jwtKeyId": <jwtKeyId>
  },
  "secrets": {
    "clientSecret": <clientSecret>,
    "privateKey": <privateKey>,
    "privateKeyPassword": <privateKeyPassword>
  }
}
```

ServiceNow ITOM OAuth
```
POST https://localhost:5601/api/actions/connector
{
  "name": "Test Servicenow ITOM OAuth API",
  "connector_type_id": ".servicenow-itom",
  "config": {
    "apiUrl": "https://dev102869.service-now.com",
    "isOAuth": true,
    "clientId": <clientId>,
    "userIdentifierValue": <email>,
    "jwtKeyId": <jwtKeyId>
  },
  "secrets": {
    "clientSecret": <clientSecret>,
    "privateKey": <privateKey>,
    "privateKeyPassword": <privateKeyPassword>
  }
}
```
</details>


3. Verify that you can create all 3 service now connector types using the API with OAuth credentials without privateKeyPassword - same requests as in (2) but exclude `privateKeyPassword` from the `secrets` config

4. Verify that you get appropriate error messages when creating all 3 connectors with missing config/secrets fields
5. Verify that you can execute all 3 service now connector types. Verify that a `connector_token` SO is created to save the access token when it is created.
  * Create the connectors via the API
  * Navigate to the Connectors page in the Kibana UI to execute the connectors

### Checklist
